vertx-web/src/main/java/io/vertx/ext/web/handler/impl/DigestAuthHandlerImpl.java

```diff
@@ -0,0 +1,228 @@
+    if (now - lastExpireRun > nonceExpireTimeout / 2) {
+      for(Iterator<Map.Entry<String,Nonce>> it = nonces.entrySet().iterator(); it.hasNext();){
+        Map.Entry<String, Nonce> entry = it.next();
+        if (entry.getValue().createdAt + nonceExpireTimeout < now) {
+          it.remove();
+        }
+      }
+
+      lastExpireRun = now;
+    }
+
+    User user = context.user();
+    if (user != null) {
+      // Already authenticated in, just authorise
+      authorise(user, context);
+    } else {
+      HttpServerRequest request = context.request();
+      String authorization = request.headers().get(HttpHeaders.AUTHORIZATION);
+
+      if (authorization == null) {
+        handle401(context);
+      } else {
+        try {
+          String sscheme;
+          int idx = authorization.indexOf(' ');
+
+          if (idx <= 0) {
+            context.fail(400);
+            return;
+          }
+
+          sscheme = authorization.substring(0, idx);
+
+          if (!"Digest".equalsIgnoreCase(sscheme)) {
+            context.fail(400);
+            return;
+          }
+
+          JsonObject authInfo = new JsonObject();
+          // Split the parameters by comma.
+          String[] tokens = authorization.substring(idx).split(",(?=(?:[^\"]|\"[^\"]*\")*$)");
+          // Parse parameters.
+          int i = 0;
+          int len = tokens.length;
+
+          while (i < len) {
+            // Strip quotes and whitespace.
+            Matcher m = PARSER.matcher(tokens[i]);
+            if (m.find()) {
+              authInfo.put(m.group(1), m.group(2));
+            }
+
+            ++i;
+          }
+
+          final String nonce = authInfo.getString("nonce");
+
+          // check for expiration
+          if (!nonces.containsKey(nonce)) {
+            handle401(context);
+            return;
+          }
+
+          // check for nonce counter (prevent replay attack
+          if (authInfo.containsKey("qop")) {
+            Integer nc = Integer.parseInt(authInfo.getString("nc"));
+            final Nonce n = nonces.get(nonce);
+            if (nc <= n.count) {
+              handle401(context);
+              return;
+            }
+            n.count = nc;
+          }
+
+          // validate the opaque value
+          final Session session = context.session();
+          if (session != null) {
+            String opaque = (String) session.data().get("opaque");
+            if (opaque != null && !opaque.equals(authInfo.getString("opaque"))) {
+              handle401(context);
+              return;
+            }
+          }
+
+          // we now need to pass some extra info
+          authInfo.put("method", context.request().method().name());
+
+          authProvider.authenticate(authInfo, res -> {
+            if (res.succeeded()) {
+              User authenticated = res.result();
+              context.setUser(authenticated);
+              authorise(authenticated, context);
+              // TODO: refresh session id
```

Comment:
once that is merged we can request regeneration of session id's which are not yet on master